File: src/net/intellij/plugins/sexyeditor/ImagePreviewPanel.java
package net.intellij.plugins.sexyeditor;

import javax.imageio.ImageIO;
import javax.swing.BorderFactory;
import javax.swing.Icon;
import javax.swing.ImageIcon;
import javax.swing.JButton;
import javax.swing.JFileChooser;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.SwingConstants;
import javax.swing.filechooser.FileNameExtensionFilter;
import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Dimension;
import java.awt.RenderingHints;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.image.BufferedImage;
import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;
import java.io.File;
import java.io.IOException;

/**
 * Accessory for JFileChooser that enables image previews.
 * Do not loads large images automatically, because of memory consumption.
 * Hides itself automatically when selecting non-images.
 */
public class ImagePreviewPanel extends JPanel implements PropertyChangeListener {

	private JLabel label;
	private JButton sure;
	private int maxImgWidth;

	private final int previewSize;
	private int maxFileSize = 512 * 1024;
	private boolean showOnlyOnImages = true;
	private String[] imageExtensions = new String[] {"gif", "jpg", "jpeg", "bmp", "png"};

	public ImagePreviewPanel() {
		this(200);
	}

	public ImagePreviewPanel(int previewSize) {
		this.previewSize = previewSize;
		setLayout(new BorderLayout(5, 5));
		setBorder(BorderFactory.createEmptyBorder(5, 5, 5, 5));
		add(new JLabel("Preview:"), BorderLayout.NORTH);
		label = new JLabel();
		label.setBackground(Color.WHITE);
		label.setOpaque(true);
		label.setHorizontalAlignment(SwingConstants.CENTER);
		label.setPreferredSize(new Dimension(previewSize, previewSize));
		maxImgWidth = previewSize - 5;
		label.setBorder(BorderFactory.createEtchedBorder());
		add(label, BorderLayout.CENTER);

		sure = new JButton("big image - click to preview");
		sure.setVisible(false);
		add(sure, BorderLayout.SOUTH);
		sure.addActionListener(e -> {
			sure.setVisible(false);
			Icon icon = loadImage();
			label.setIcon(icon);
			ImagePreviewPanel.this.repaint();
		});
	}

	/**
	 * Attach this accessory to file chooser and, optionally, add file filter for image files.
	 */
	public void attachToFileChooser(JFileChooser fileChooser, String imagesFilterName) {
		fileChooser.setAccessory(this);
		fileChooser.addPropertyChangeListener(this);
		if (imagesFilterName != null) {
			fileChooser.setFileFilter(new FileNameExtensionFilter(imagesFilterName, imageExtensions));
		}
	}

	private File newFile;

	public void propertyChange(PropertyChangeEvent evt) {
		Icon icon = null;
		if (JFileChooser.SELECTED_FILE_CHANGED_PROPERTY.equals(evt.getPropertyName())) {
			if (showOnlyOnImages) {
				this.setVisible(false);
			}
			newFile = (File) evt.getNewValue();
			if (newFile != null) {
				String path = newFile.getAbsolutePath();
				int i = path.lastIndexOf('.');
				if (i != -1) {
					String extension = path.substring(i + 1).toLowerCase();
					for (String ext : imageExtensions) {
						if (extension.equals(ext)) {
							if (showOnlyOnImages) {
								this.setVisible(true);
							}
							sure.setVisible((maxFileSize != -1) && (newFile.length() > maxFileSize));
							if (!sure.isVisible()) {
								icon = loadImage();
							}
							break;
						}
					}
				}
			}
			label.setIcon(icon);
			this.repaint();
		}
	}

	/**
	 * Loads and scales images for previewing purposes if necessary.
	 */
	private Icon loadImage() {
		if (newFile == null) {
			return null;
		}
		try {
			BufferedImage image = ImageIO.read(newFile);
			float width = image.getWidth();
			float height = image.getHeight();
			if (width <= maxImgWidth) {
				return new ImageIcon(image);
			}
			float scale = height / width;
			width = maxImgWidth;
			height = (width * scale); // height should be scaled from new width
			return new ImageIcon(SwingUtil.getScaledInstance(image, Math.max(1, (int) width), Math.max(1, (int) height), RenderingHints.VALUE_INTERPOLATION_BILINEAR, true));
		}
		catch (IOException ioex) {
			// ignore, couldn't read image
		}
		finally {
			newFile = null;
		}
		return null;
	}

	// ---------------------------------------------------------------- accessories

	/**
	 * Returns preview size.
	 */
	public int getPreviewSize() {
		return previewSize;
	}

	public int getMaxFileSize() {
		return maxFileSize;
	}

	/**
	 * Sets maximal file size of image that will be previewed automatically.
	 * By setting this value to <code>-1</code> all images will be previewed.  
	 */
	public void setMaxFileSize(int maxFileSize) {
		this.maxFileSize = maxFileSize;
	}

	public boolean isShowOnlyOnImages() {
		return showOnlyOnImages;
	}

	public void setShowOnlyOnImages(boolean showOnlyOnImages) {
		this.showOnlyOnImages = showOnlyOnImages;
	}
}

File: src/main/kotlin/com/github/igr/sexyeditor/ui/ImagePreviewPanel.kt
package com.github.igr.sexyeditor.ui

import com.intellij.ui.JBColor
import java.awt.BorderLayout
import java.awt.Dimension
import java.awt.RenderingHints
import java.awt.event.ActionEvent
import java.awt.image.BufferedImage
import java.beans.PropertyChangeEvent
import java.beans.PropertyChangeListener
import java.io.File
import java.io.IOException
import java.util.*
import javax.imageio.ImageIO
import javax.swing.*
import javax.swing.filechooser.FileNameExtensionFilter
import kotlin.math.max

/**
 * @param previewSize - size of preview in pixels
 * @param maxFileSize - maximum file size in pixels to automatically show preview for. Set  `-1` to remove limitation
 */
class ImagePreviewPanel(
    private val previewSize: Int = 200,
    private val maxFileSize : Int = 512 * 1024,
) : JPanel(), PropertyChangeListener {

    private val label: JLabel
    private val sure: JButton
    private var maxImgWidth = 0
    private val showOnlyOnImages = false
    private val imageExtensions = arrayOf("gif", "jpg", "jpeg", "bmp", "png")

    init {
        layout = BorderLayout(5, 5)
        border = BorderFactory.createEmptyBorder(5, 5, 5, 5)
        add(JLabel("Preview:"), BorderLayout.NORTH)

        label = JLabel().apply {
            background = JBColor.WHITE
            isOpaque = true
            horizontalAlignment = SwingConstants.CENTER
            preferredSize = Dimension(previewSize, previewSize)
            border = BorderFactory.createEtchedBorder()
        }

        maxImgWidth = previewSize - 5
        add(label, BorderLayout.CENTER)
        sure = JButton("big image - click to preview").apply {
            isVisible = false
        }
        add(sure, BorderLayout.SOUTH)
        sure.addActionListener { e: ActionEvent? ->
            sure.isVisible = false
            val icon: Icon? = loadImage()
            label.icon = icon
            this@ImagePreviewPanel.repaint()
        }
    }

    /**
     * Attach this accessory to file chooser and, optionally, add file filter for image files.
     */
    fun attachToFileChooser(fileChooser: JFileChooser, imagesFilterName: String?) {
        fileChooser.accessory = this
        fileChooser.addPropertyChangeListener(this)
        if (imagesFilterName != null) {
            fileChooser.fileFilter = FileNameExtensionFilter(imagesFilterName, *imageExtensions)
        }
    }

    private var newFile: File? = null

    override fun propertyChange(evt: PropertyChangeEvent) {
        var icon: Icon? = null
        if (JFileChooser.SELECTED_FILE_CHANGED_PROPERTY == evt.propertyName) {
            if (showOnlyOnImages) {
                this.isVisible = false
            }
            if (evt.newValue != null) {
                newFile = evt.newValue as File
                val path: String = newFile!!.absolutePath
                val i = path.lastIndexOf('.')
                if (i != -1) {
                    val extension = path.substring(i + 1).lowercase(Locale.getDefault())
                    for (ext in imageExtensions) {
                        if (extension == ext) {
                            if (showOnlyOnImages) {
                                this.isVisible = true
                            }
                            sure.isVisible = maxFileSize != -1 && newFile!!.length() > maxFileSize
                            if (!sure.isVisible) {
                                icon = loadImage()
                            }
                            break
                        }
                    }
                }
            }
            label.icon = icon
            this.repaint()
        }
    }

    /**
     * Loads and scales images for previewing purposes if necessary.
     */
    private fun loadImage(): Icon? {
        if (newFile == null) {
            return null
        }
        try {
            val image: BufferedImage = ImageIO.read(newFile)
            var width = image.width.toFloat()
            var height = image.height.toFloat()
            if (width <= maxImgWidth) {
                return ImageIcon(image)
            }
            val scale = height / width
            width = maxImgWidth.toFloat()
            height = width * scale // height should be scaled from new width
            return ImageIcon(
                createScaledInstance(
                    image,
                    max(1, width.toInt()),
                    max(1, height.toInt()),
                    RenderingHints.VALUE_INTERPOLATION_BILINEAR,
                    true
                )
            )
        } catch (ignore: IOException) {
        } finally {
            newFile = null
        }
        return null
    }
}